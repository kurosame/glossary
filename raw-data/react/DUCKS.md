## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
