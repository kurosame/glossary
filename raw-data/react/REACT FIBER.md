## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
