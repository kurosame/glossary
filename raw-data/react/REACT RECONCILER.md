## category

react

## titles

React Reconciler
react-reconciler

## description

VDOM と呼ばれている部分

1. ReactElement のツリーを構築
1. Component をマウント、アンマウント、更新
1. ライフサイクル関数を呼出す
1. 現在の DOM との差分を計算する
1. renderer に差分を伝える
