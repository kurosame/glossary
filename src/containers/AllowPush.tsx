import React from 'react'
import Snackbar from '@material-ui/core/Snackbar'
import Button from '@material-ui/core/Button'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import { requestPermission, isPermission } from '@/utils/messaging'

const AllowPush: React.FC = () => {
  const [open, setOpen] = React.useState(isPermission() === 'default')
  const handleRequestPermission = (): void => requestPermission()
  const handleClose = (): void => setOpen(false)

  return (
    <Snackbar
      anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
      open={open}
      message="Can test push notifications"
      action={
        <>
          <Button
            color="primary"
            data-testid="allow-push-ok"
            onClick={(): void => {
              handleRequestPermission()
              handleClose()
            }}
          >
            OK
          </Button>
          <IconButton color="inherit" data-testid="allow-push-close" onClick={handleClose}>
            <CloseIcon />
          </IconButton>
        </>
      }
    />
  )
}

export default AllowPush
