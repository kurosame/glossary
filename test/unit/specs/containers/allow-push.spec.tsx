/**
 * @jest-environment jsdom
 */

import React from 'react'
import { render, fireEvent, cleanup, RenderResult } from '@testing-library/react'
import AllowPush from '@/containers/AllowPush'

const requestPermission = jest.fn()

// jest.mock runs at top level
jest.mock('@/utils/messaging', () => ({
  requestPermission: (): jest.Mock => requestPermission,
  isPermission: (): string => 'default'
}))

let wrapper: RenderResult
beforeEach(() => {
  wrapper = render(<AllowPush />)
})
afterEach(cleanup)

test('Click `allow-push-ok` will call `requestPermission` and close snack bar', () => {
  expect(wrapper.container.querySelector('div')).not.toBeNull()

  fireEvent.click(wrapper.getByTestId('allow-push-ok'))

  setTimeout(() => {
    expect(requestPermission).toBeCalled()
    expect(wrapper.container.querySelector('div')).toBeNull()
  }, 1000)
})

test('Click `allow-push-close` will close snack bar', () => {
  expect(wrapper.container.querySelector('div')).not.toBeNull()

  fireEvent.click(wrapper.getByTestId('allow-push-close'))

  setTimeout(() => {
    expect(requestPermission).not.toBeCalled()
    expect(wrapper.container.querySelector('div')).toBeNull()
  }, 1000)
})

test('Match the snapshot', () => {
  expect(wrapper.asFragment()).toMatchSnapshot()
})
