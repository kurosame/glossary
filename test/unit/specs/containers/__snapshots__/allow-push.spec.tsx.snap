// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomRight"
  >
    <div
      class="MuiPaper-root MuiSnackbarContent-root MuiPaper-elevation6"
      direction="up"
      role="alert"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message"
      >
        Can test push notifications
      </div>
      <div
        class="MuiSnackbarContent-action"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
          data-testid="allow-push-ok"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            OK
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          data-testid="allow-push-close"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
